### DISCUSSION TITLE
Lightweight printing from Linux based embedded system

### DISCUSSION THREAD

**Discussion by qo0op** [OpenPrinting/cups#307] [Q&A] [answered]
Hi,

I would be interested in your opinion what would be the best lightweight solution for printing from Linux based embedded system (e.g. BeagleBone, ...) these days?

General requirements/idea is to have: 
- small memory footprint
- possibility to use printers connected over USB
  - network, Wifi direct support is nice to have
  - legacy printers don't have to be supported
  - support for wide spectrum of new printers if possible
- print from command line (using 'lp', 'lpr'), Qt application

From what I was able to find so far "cups + ipp-usb" could be the 'minimal' solution . Do I need "cups-filters" in case I would use some document supported by printer directly (e.g. JPEG)? (Then I don't need any 'filters' and 'backend' is supplied by 'ipp-usb', right?.)

Does it make sense to think about using [HPLIP](https://layers.openembedded.org/layerindex/recipe/826/) for direct usb printing rather for HP printers?

Thank you in advance!

**Comment by zdohnal** [comment, marked as answer]
Hi @qo0op ,

IMO cups+ipp-usb+cups-filters is the minimal solution - cups-filters in general provides conversion between more types than CUPS provides by default, so its inclusion prevents possible issues for users when they send an unexpected document format into CUPS alone.

AFAIK ipp-usb doesn't provide any filters or backends - its daemon creates an interface, which is used by CUPS for communication with IPP-over-USB capable printer. So ipp-usb is needed.

HPLIP is not needed unless you want to support legacy printers or you want to have an alternative if driverless printing doesn't work.
